circuit-prover/src/batch_stark_prover.rs

```diff
@@ -0,0 +1,669 @@
+//! Batch STARK prover and verifier that unifies all circuit tables
+//! into a single multi-proof using `p3-multi-stark`.
+
+use alloc::string::String;
+use alloc::vec;
+use alloc::vec::Vec;
+
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_circuit::tables::Traces;
+use p3_field::{BasedVectorSpace, Field};
+use p3_matrix::Matrix;
+use p3_matrix::dense::RowMajorMatrix;
+use p3_mmcs_air::air::{MmcsTableConfig, MmcsVerifyAir};
+use p3_multi_stark::{MultiProof, StarkGenericConfig as MSGC, StarkInstance, Val as MVal};
+use thiserror::Error;
+use tracing::instrument;
+
+use crate::air::{AddAir, ConstAir, MulAir, PublicAir, WitnessAir};
+use crate::config::StarkField;
+use crate::field_params::ExtractBinomialW;
+use crate::prover::TablePacking;
+
+#[repr(usize)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Table {
+    Witness = 0,
+    Const = 1,
+    Public = 2,
+    Add = 3,
+    Mul = 4,
+    Mmcs = 5,
+}
+
+/// Number of circuit tables included in the unified multi‑STARK proof.
+pub const NUM_TABLES: usize = Table::Mmcs as usize + 1;
+
+/// Row counts wrapper with type-safe indexing.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct RowCounts([usize; NUM_TABLES]);
+
+impl RowCounts {
+    /// Creates a new RowCounts with the given row counts for each table.
+    pub fn new(rows: [usize; NUM_TABLES]) -> Self {
+        // Validate that all row counts are non-zero
+        for (i, &count) in rows.iter().enumerate() {
+            assert!(count > 0, "Table {i} has zero rows, which is not supported");
+        }
+        Self(rows)
+    }
```

Comment:
I think we can make this function constant

```suggestion
    pub const fn new(rows: [usize; NUM_TABLES]) -> Self {
        let mut i = 0;
        // Validate that all row counts are non-zero
        while i < NUM_TABLES {
            if rows[i] == 0 {
                panic!("Table {i} has zero rows");
            }
            i += 1;
        }
        Self(rows)
    }
```